lib/grell/crawler.rb

```diff
@@ -51,12 +51,21 @@ def crawl(site, block)
       Grell.logger.info "Visiting #{site.url}, visited_links: #{@collection.visited_pages.size}, discovered #{@collection.discovered_pages.size}"
       site.navigate
       filter!(site.links)
-
-      if block #The user of this block can send us a :retry to retry accessing the page
-        while block.call(site) == :retry
-          Grell.logger.info "Retrying our visit to #{site.url}"
-          site.navigate
-          filter!(site.links)
+      add_redirect_url(site)
+
+      if block # The user of this block can send us a :retry to retry accessing the page
+        begin
+          while block.call(site) == :retry
+            Grell.logger.info "Retrying our visit to #{site.url}"
+            site.navigate
+            filter!(site.links)
+            add_redirect_url(site)
+          end
+        rescue Capybara::Poltergeist::BrowserError, Capybara::Poltergeist::DeadClient,
```

Comment:
The only code that can raise all this is site.navigate that already catch exceptions itself (I think you found a case when that was not true but looking at the code it does not seem possible)